File: Implementations/Recursion/2_Sort_array.cpp
/*
Author : humanbeeng
*/
#include <bits/stdc++.h>
using namespace std;

void insert_proper(vector<int> &arr, int element) {
    if (arr.size() == 0 || arr[arr.size() - 1] <= element) {
        arr.push_back(element);
        return;
    }

    int temp = arr[arr.size() - 1];
    arr.pop_back();
    insert_proper(arr, element);
    arr.push_back(temp);
}

void sort_arr(vector<int> &arr) {
    if (arr.size() == 1 || arr.size() == 0) {
        return;
    }
    int curr_top = arr[arr.size() - 1];
    arr.pop_back();

    sort_arr(arr);

    insert_proper(arr, curr_top);
    return;
}

int main() {
    int n_elements;
    cin >> n_elements;

    vector<int> arr;

    for (int i = 0; i < n_elements; i++) {
        int num;
        cin >> num;

        arr.push_back(num);
    }

    for (int num : arr) {
        cout << num << " ";
    }
    cout << '\n';

    sort_arr(arr);
    cout << '\n';
    for (int num : arr) {
        cout << num << " ";
    }
    cout << '\n';

    return 0;
}

File: Implementations/Recursion/2_Sort_array.py


def _insert_proper(arr, element):
    if(len(arr) == 0 or arr[len(arr) - 1] <= element):
        arr.append(element)
        return

    temp = int(arr[len(arr) - 1])
    arr.pop(len(arr) - 1)
    _insert_proper(arr, element)
    arr.append(temp)
    return


def _sort_arr(arr):
    n_elements = int(len(arr))
    if(n_elements == 0 or n_elements == 1):
        return
    curr_top = int(arr[n_elements - 1])
    arr.pop(n_elements - 1)
    _sort_arr(arr)
    _insert_proper(arr, curr_top)
    return


if __name__ in "__main__":
    n_elements = int(input())
    arr = list(map(int, input().split()))
    print("Before sorting the array")
    print(arr)

    _sort_arr(arr)

    print("After sorting the array")
    print(arr)
